File: 2022/Day16/main.cpp
#include <fstream>
#include <iostream>
#include <string>
#include <string_view>
#include <vector>
#include <unordered_map>
#include <optional>
#include <ranges>
#include <memory>
#include <concepts>
#include <algorithm>
#include <numeric>
#include <array>

constexpr int MINUTES = 30;

std::vector<std::string> read_file(std::string_view filename) {
    std::ifstream file(filename.data());
    std::vector<std::string> lines;
    std::string line;
    while (std::getline(file, line)) {
        lines.push_back(line);
    }
    return lines;
}

enum class State : bool {
    OPEN,
    CLOSED
};

namespace v = std::ranges::views;
namespace r = std::ranges;
using namespace std::string_view_literals;
struct Valve;


template <typename T>
concept string_like = std::same_as<T, std::string> || std::same_as<T, std::string_view>;

struct HashTo {
    using is_transparent = void;
    std::size_t operator()(const string_like auto& o) const {
        using T = std::remove_cvref_t<decltype(o)>;
        return std::hash<T>{}(o);
    }
};

struct EqualTo {
    using is_transparent = void;
    constexpr bool operator()(const string_like auto& lhs,  const string_like auto& rhs) const {
        return std::string_view{lhs} == std::string_view{rhs};
    }
};


using ValveMap = std::unordered_map<std::string, std::shared_ptr<Valve>, HashTo, EqualTo>;

struct Valve {
    std::string name;
    std::optional<int> _flow_rate;
    std::array<int, MINUTES> flow_rate_minutes{};
    ValveMap tunnels;
    State state;

    template <typename T>
    using opt = std::optional<T>;
    static constexpr auto to_str = [](auto &&subr) { return std::string{subr.begin(), subr.end()}; };

    constexpr static auto none = std::nullopt;

    Valve(const Valve& other) : name(other.name), _flow_rate(other._flow_rate), flow_rate_minutes(other.flow_rate_minutes), tunnels(other.tunnels), state(other.state) {

    }

    Valve(Valve&& other) : name(std::move(other.name)), _flow_rate(std::move(other._flow_rate)), flow_rate_minutes(other.flow_rate_minutes), tunnels(std::move(other.tunnels)), state(other.state) {

    }

    Valve(std::string_view _name, opt<int> flow_rate = none, opt<std::string_view> tunnelstr = none, ValveMap* valves = nullptr) 
    : name(_name), _flow_rate(std::move(flow_rate)), tunnels{}
    {
        this->state = State::CLOSED;
        if (tunnelstr.has_value() && valves != nullptr) {
            set_tunnels(tunnelstr.value(), *valves);
        }
    }

    int flow_rate() const {
        if (this->_flow_rate.has_value()) {
            return this->_flow_rate.value();
        }
        return 0;
    }

    int flow_rate(int i) const {
        if (this->_flow_rate.has_value()) {
            return this->flow_rate_minutes[i];
        }
        return 0;
    }

    void set_flow_rate(int flow_rate) {
        this->_flow_rate = flow_rate;
        if (flow_rate > 0) {
            for (int i = 0; i < MINUTES; i++) {
                this->flow_rate_minutes[i] = flow_rate * (MINUTES - i);
            }
        }
    }

    void set_tunnels(std::string_view tunnelstr, ValveMap& valves) {
        for (const auto& tunnel : v::split(tunnelstr, ", "sv) | v::transform(to_str)) {
            auto it = valves.find(tunnel);
            if (it == valves.end()) {
                it = valves.insert({tunnel, std::make_shared<Valve>(tunnel)}).first;
            }
            this->tunnels.insert(*it);
        }
    }
};


int main() {
    auto lines = read_file("sample_data.txt");
    ValveMap valves;
    for (const auto& line : lines) {
        char name[256];
        int flow_rate = 0;
        int ptr;
        sscanf(line.data(), "Valve %s has flow rate=%d; tunnels lead to valves %n", name, &flow_rate, &ptr);
        std::string sname{name};
        std::string_view tunnelstr{line.data() + ptr};
        if (auto it = valves.find(sname); it != valves.end()) {
            it->second->set_flow_rate(flow_rate);
            it->second->set_tunnels(tunnelstr, valves);
        } else {
            valves.insert({sname, std::make_shared<Valve>(sname, flow_rate, tunnelstr, &valves)});
        }
    }
    auto current_valve = valves["AA"];
    int tot_pressure = 0;
    for (int i = 0; i < MINUTES; ++i) {
        if (current_valve->flow_rate() > 0 && current_valve->state == State::CLOSED) {
            current_valve->state = State::OPEN;
            tot_pressure += current_valve->flow_rate(i);
            std::cout << "Valve " << current_valve->name << " is open at minute " << (i+1) << '\n';
        } else {
            auto filter = v::transform(current_valve->tunnels, [](const auto& v){ return v.second; }) | v::filter([](const auto& v) { return v->state == State::CLOSED; });
            current_valve = r::max(filter, {}, [i](const auto& v) { return v->flow_rate(i); });
            std::cout << "Moving to " << current_valve->name << " at minute " << (i+1) << '\n';
        }
    }
    std::cout << "Total pressure: " << tot_pressure << '\n';
    return 0;
}

File: 2022/Day16/main.py
from __future__ import annotations
import re
import numpy
from enum import IntEnum

patt = re.compile(r'Valve (\w+) has flow rate=(\d+); tunnels? leads? to valves? ((?:\w+(?:, )?)+)', re.MULTILINE)

SAMPLE_DATA: bool = True
filename = "sample_data.txt" if SAMPLE_DATA else "data.txt"
with open(filename, 'r') as f:
    data = f.read()

MAX_MINUTES = 30

class State(IntEnum):
    OPEN = 1
    CLOSED = 2

class Valve:
    name: str
    _flow_rate: None|int
    time_flow_rate: numpy.ndarray
    tunnels: list[Valve]
    state: State

    def __init__(self, name: str, flow_rate: str|None = None, tunnels: str|None = None, valves: dict[str, Valve]|None = None):
        self.name = name
        self._flow_rate = int(flow_rate) if flow_rate else None
        self.tunnels = []
        self.state = State.CLOSED
        if tunnels and (valves is not None):
            self.set_tunnels(tunnels, valves)

    def can_open(self):
        return self.flow_rate > 0 and self.state == State.CLOSED
    
    def set_tunnels(self, tunnels: str, valves: dict[str, Valve]):
        connected_tunnels = tunnels.split(', ')
        for tunnel in connected_tunnels:
            if tunnel not in valves:
                valves[tunnel] = Valve(tunnel)
        self.tunnels = [valves[tunnel] for tunnel in connected_tunnels]
    
    @property
    def flow_rate(self):
        return self._flow_rate if self._flow_rate else 0
    
    @flow_rate.setter
    def flow_rate(self, value: int|str):
        self._flow_rate = value if isinstance(value, int) else int(value)
        if self._flow_rate > 0:
            self.time_flow_rate = numpy.array([i*self._flow_rate for i in range(MAX_MINUTES, 0, -1)], dtype=int)

    def flow_rate_at(self, minute: int):
        return self.time_flow_rate[minute - 1] if self._flow_rate else 0

    def __repr__(self):
        return f"Valve {self.name} has flow rate={self.flow_rate}; tunnels lead to valves {', '.join([valve.name for valve in self.tunnels])}"
    
    def __str__(self):
        return self.__repr__()

def print_dict(d: dict[str, Valve]):
    for key, value in d.items():
        print(f"{key}: {value}")


tot_pressure = 0
valves: dict[str, Valve] = {}
for match in patt.finditer(data):
    name, flow_rate, tunnels = match.groups()
    if valve := valves.get(name):
        valve.flow_rate = flow_rate
        valve.set_tunnels(tunnels, valves)
    else:
        valves[name] = Valve(name, flow_rate, tunnels, valves)

max_pressure = 0

class Move(IntEnum):
    MOVE = 1
    OPEN = 2

def format_tuple(t: tuple[Valve, Move]):
    return f"{'o' if t[1] == Move.OPEN else 'm'}{t[0].name}"

def update_pressure(pressure: int):
    global max_pressure
    if pressure > max_pressure:
        print(f'New max pressure: {pressure}')
        max_pressure = pressure

def visit(current_valve: Valve, pressure: int, minute: int):
    if minute >= MAX_MINUTES:
        update_pressure(pressure)
        return
    opened = False
    if current_valve.can_open():
        current_valve.state = State.OPEN
        minute += 1
        pressure += current_valve.flow_rate_at(minute)
        opened = True
    for valve in current_valve.tunnels:
        if valve.state == State.CLOSED:
            visit(valve, pressure, minute+1)
    if opened:
        minute -= 1
    current_valve.state = State.CLOSED
    for valve in current_valve.tunnels:
        if valve.state == State.CLOSED:
            visit(valve, pressure, minute+1)
    return


current_valve = valves['AA']
visit(current_valve, 1, 0)
print(max_pressure)